---
title: "RNA-Seq Exploratory Data Analysis"
author: "Kaylene"
format:
  html:
    self-contained: true
  pdf:
    toc: true
    number-sections: true
    include-in-header:
      text: |
        \usepackage{fvextra}
        \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
        \DefineVerbatimEnvironment{OutputCode}{Verbatim}{breaklines,commandchars=\\\{\}}
editor: 
  markdown: 
    wrap: 72
---

```{r}
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(patchwork)
library(scales)
library(ComplexHeatmap)
library(EnhancedVolcano)
library(DESeq2)
```

# Prepare Data

```{r}
ccle_counts <- read.csv("../data/processed/ccle_counts.csv", row.names = 1)
ccle_meta <- read.csv("../data/processed/ccle_meta.csv")
```

```{r}
identical(colnames(ccle_counts), ccle_meta$CCLE_ID)
## the column names of ccle_counts and the content of the CCLE_ID column in ccle_meta are equivalent, both in terms of content and order of contents
```

```{r}
## subset only the stomach cancer cell lines in the meta data and the counts matrix 
ccle_meta_subset <- ccle_meta[ccle_meta$Site_Primary == "stomach",]
ccle_counts_subset <- ccle_counts[,ccle_meta_subset$CCLE_ID]
```

```{r}
## data normalization
dds <- DESeqDataSetFromMatrix(countData = data.frame(ccle_counts_subset),
                              colData = ccle_meta_subset,
                              design = ~ Pathology) ## Pathology is just a dummy variable until we define clusters
dim(dds)
```

```{r}
dds <- dds[rowSums(counts(dds)) >= 100,]
vsd <- vst(dds) ## variance stabilized counts
dim(dds)
```

```{r}
print(17668 - 16196) ## 1472 genes removed when removing extremely lowly expressed genes
```

```{r}
head(ccle_meta_subset)
counts(dds)[1:10,1:10]
assay(vsd)[1:10,1:10]
```

# Principal Component Analysis
```{r}
unique(ccle_meta_subset$tcga_code) ## only one tcga code so can't test clustering that way
```

```{r}
plotPCA(vsd, intgroup = "Pathology", ntop = 500) ## no obvious clustering by pathology
plotPCA(vsd, intgroup = "Gender", ntop = 500) ## no obvious clustering by gender
plotPCA(vsd, intgroup = c("Gender", "Pathology"), ntop = 500) ## no obvious clustering by both pathology and gender
```

```{r}
variances <- rowVars(assay(vsd)) # calculate variance across samples of all genes
top_var_genes <- order(variances, decreasing = TRUE)[1:500] # extract the indices of the top 500 variable genes in descending order
high_var_mat <- assay(vsd)[top_var_genes,] # subset the gene counts with the top 500 variable genes
pca_res <- prcomp(t(high_var_mat), center = TRUE, scale = FALSE) # do PCA
explained_variance <- pca_res$sdev^2 / sum(pca_res$sdev^2) * 100 # create a vector with the % of variance explained by each PC
```

PC3 and PC4 do not show as obvious clustering/grouping as PC1 and PC2

```{r}
df_pca_3_4 <- data.frame(PC3 = pca_res$x[,3], PC4 = pca_res$x[,4], ccle_meta_subset)
df_pca_3_4 |> ggplot() + aes(x = PC3, y = PC4, color = Gender) +
  geom_point(size = 3, alpha = 0.5) + 
  theme_classic() + 
  labs(x = paste0("PC3: ", round(explained_variance[3], 1), "% variance"), 
       y = paste0("PC4: ", round(explained_variance[4], 1), "% variance"))
```

The below elbow plot displays the principal components based on the
percentage of the total variance each one captures. Principle Component
1 and Principle Component 2 capture the largest portion of variance in
the data by far, so examining the other PCs would only be marginally
beneficial.

```{r}
df_variance <- data.frame(PC = seq_len(25), Variance = explained_variance[1:25])
ggplot(df_variance, aes(x = PC, y = Variance)) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(title = "Elbow Plot (Top 25 PCs) for stomach cancer lines in CCLE", x = "Principal Component", y = "Explained Variance (%)")
```

Analysis of highest load genes in PC1

```{r}
## determine the genes with the highest loadings (contributions/influence) in principle component 1
pc1_loadings <- pca_res$rotation[, 1]
pc1_sorted_loadings <- sort(pc1_loadings, decreasing = TRUE)
head(pc1_sorted_loadings, n = 6)
```

```{r}
# extract the gene expression data for the genes with the highest loadings in PC1
top_6_pc1_genes <- names(head(pc1_sorted_loadings, n = 6))
vsd_top_6_genes_PC1 <- assay(vsd)[top_6_pc1_genes,]
```

```{r}
## reshape vsd_top_6_genes_PC1 so that the gene names are in the first column, and each gene name is repeated for the total number of samples
vsd_top_6_longer_PC1 <- as.data.frame(vsd_top_6_genes_PC1) |> mutate(Gene = rownames(vsd_top_6_genes_PC1)) |> pivot_longer(cols = -Gene, names_to = "Sample", values_to = "Expression")
```

```{r}
ggplot(vsd_top_6_longer_PC1, aes(x = Gene, y = Expression)) +
  geom_boxplot(aes(fill = Gene), alpha = 0.5) +
  geom_jitter(width = 0.2, size = 1, alpha = 0.5) +
  scale_y_continuous(limits = c(0, 20)) +
  labs(title = "Expression of Top PC1 Genes",
       x = "Gene",
       y = "Expression (VST)") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r}
df_pca <- data.frame(PC1 = pca_res$x[, 1], PC2 = pca_res$x[, 2])
```

```{r}
df_facet_PC1 <- df_pca
for (gene in top_6_pc1_genes){
  df_facet_PC1[[gene]] <- assay(vsd)[gene, ]
}
```

```{r}
df_facet_longer_PC1 <- df_facet_PC1 |>
  mutate(Sample = rownames(df_facet_PC1)) |>      # make row names a column
  pivot_longer(cols = LYZ:VIL1,               # select gene columns
               names_to = "Gene",
               values_to = "Expression")
```

```{r}
df_facet_longer_PC1 |>
  ggplot(aes(x = PC1, y = PC2, color = Expression)) +
  geom_point(size = 1.5, alpha = 0.8) +
  facet_wrap(~ Gene, nrow = 2) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_classic() +
  labs(
    title = "PC1 vs PC2 Colored by Gene Expression (Top 6 Genes in PC1)",
    x = "PC1",
    y = "PC2",
    color = "Expression Value"
  )
```

The top 6 genes with the highest loadings/contributions in PC1:

LYZ – Lysozyme, an antimicrobial enzyme often expressed in monocytes,
macrophages, and some epithelial cells. AGR2 – Anterior Gradient 2,
associated with mucus production and epithelial differentiation, often
upregulated in gastrointestinal tissues and some cancers. LGALS4 –
Galectin-4, involved in cell adhesion and signaling, typically expressed
in intestinal epithelial cells. TSPAN8 – Tetraspanin 8, plays a role in
cell migration, metastasis, and epithelial biology. GPX2 – Glutathione
Peroxidase 2, an antioxidant enzyme mainly found in the gastrointestinal
tract. VIL1 – Villin 1, an actin-binding protein highly expressed in
intestinal epithelial cells and a classic enterocyte marker.

The genes LYZ, AGR2, LGALS4, TSPAN8, GPX2, and VIL1 encode proteins with
diverse functions, most notably linked to the gastrointestinal tract
epithelium and the development and prognosis of various epithelial
cancers, particularly colorectal cancer (CRC).

Analysis for highest load genes for PC2

```{r}
## determine the genes with the highest loadings (contributions/influence) in principle component 2
pc2_loadings <- pca_res$rotation[, 2]
pc2_sorted_loadings <- sort(pc2_loadings, decreasing = TRUE)
head(pc2_sorted_loadings, n = 6)
```

```{r}
# extract the gene expression data for the genes with the highest loadings in PC2
top_6_pc2_genes <- names(head(pc2_sorted_loadings, n = 6))
vsd_top_6_genes_PC2 <- assay(vsd)[top_6_pc2_genes,]
```

```{r}
## reshape vsd_top_6_genes so that the gene names are in the first column, and each gene name is repeated for the total number of samples
vsd_top_6_longer_PC2 <- as.data.frame(vsd_top_6_genes_PC2) |> mutate(Gene = rownames(vsd_top_6_genes_PC2)) |> pivot_longer(cols = -Gene, names_to = "Sample", values_to = "Expression")
```

```{r}
ggplot(vsd_top_6_longer_PC2, aes(x = Gene, y = Expression)) +
  geom_boxplot(aes(fill = Gene), alpha = 0.5) +
  geom_jitter(width = 0.2, size = 1, alpha = 0.5) +
  scale_y_continuous(limits = c(0, 20)) +
  labs(title = "Expression of Top PC2 Genes",
       x = "Gene",
       y = "Expression (VST)") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r}
df_facet_PC2 <- df_pca
for (gene in top_6_pc2_genes){
  df_facet_PC2[[gene]] <- assay(vsd)[gene, ]
}
```

```{r}
df_facet_longer_PC2 <- df_facet_PC2 |>
  mutate(Sample = rownames(df_facet_PC2)) |>      # make row names a column
  pivot_longer(cols = ALPP:LAMB3,               # select gene columns
               names_to = "Gene",
               values_to = "Expression")
```

```{r}
df_facet_longer_PC2 |>
  ggplot(aes(x = PC1, y = PC2, color = Expression)) +
  geom_point(size = 1.5, alpha = 0.8) +
  facet_wrap(~ Gene, nrow = 2) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_classic() +
  labs(
    title = "PC1 vs PC2 Colored by Gene Expression (Top 6 Genes in PC2)",
    x = "PC1",
    y = "PC2",
    color = "Expression Value"
  )
```

ALPP, TACSTD2, LAMC2, KLK6, MUC16, and LAMB3 encode various proteins,
most of which are associated with cell adhesion, the extracellular
matrix, and are frequently implicated in cancer development and
metastasis.

# Hierarchical Clustering

```{r, fig.height=10, fig.width=8}
high_var_genes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 50) ## extract the top 50 genes with the most variation
high_var_mat <- assay(vsd)[high_var_genes,] ## extract the expression data for the top 50 variable genes

hc <- hclust(dist(t(high_var_mat)), method = "complete") ## create a dendrogram for the top 50 variable genes
clusters <- cutree(hc, k = 2) ## cut the dendrogram into 2 clusters

sample_annotation_df <- ccle_meta_subset
sample_annotation_df <- sample_annotation_df %>% dplyr::select(Pathology, Gender) ## select the pathology and gender columns from the metadata

sample_annotation_colors <- list(
  Pathology = c("primary" = "black", "metastasis" = "grey"),
  Gender = c("male" = "#009E73", "female" = "#F0E442", "missing_info" = "#D55E00")
)

column_annotation <- HeatmapAnnotation(
  df = sample_annotation_df,
  col = sample_annotation_colors,
  annotation_label = c("Pathology", "Gender")
)

Heatmap(high_var_mat,
        clustering_distance_columns = "euclidean", ## compute distance between samples with euclidean metric
        clustering_method_columns = "complete", ## use complete linkage as a clustering method
        clustering_distance_rows = "euclidean",
        clustering_method_rows = "complete",
        show_row_names = TRUE,
        show_column_names = FALSE,
        name = "VST_Counts",
        column_split = clusters, ## split the heatmap by the calculated clusters
        top_annotation = column_annotation) 
```

No clear evidence that there is a clustering that corresponds with the
metadata.

```{r}
## top 50 variable genes 
variances <- rowVars(assay(vsd)) 

top_var_genes <- order(variances, decreasing = TRUE)[1:50]

high_var_mat <- assay(vsd)[top_var_genes,]

pca_res <- prcomp(t(high_var_mat), center = TRUE, scale = FALSE) 

explained_variance <- pca_res$sdev^2 / sum(pca_res$sdev^2) * 100 ## vector of length = # of PCs where each value is the standard deviation of that PC

hc <- hclust(dist(t(high_var_mat)), method = "complete")
clusters <- cutree(hc, k = 2) ## 2 clusters

df_pca <- data.frame(PC1 = pca_res$x[, 1], PC2 = pca_res$x[, 2], Cluster = factor(clusters)) ## rows are samples


ggplot(df_pca, aes(x = PC1, y = PC2, color = Cluster)) +
  geom_point(size = 3, alpha = 0.5) + 
  scale_color_manual(values = hue_pal()(length(unique(clusters)))) +
  theme_classic() + 
  labs(x = paste0("PC1: ", round(explained_variance[1], 1), "% variance"), 
       y = paste0("PC2: ", round(explained_variance[2], 1), "% variance"))
```

```{r}
# top 500 variable genes

variances <- rowVars(assay(vsd)) 

top_var_genes <- order(variances, decreasing = TRUE)[1:500]

high_var_mat <- assay(vsd)[top_var_genes,]

pca_res <- prcomp(t(high_var_mat), center = TRUE, scale = FALSE) 

explained_variance <- pca_res$sdev^2 / sum(pca_res$sdev^2) * 100 

hc <- hclust(dist(t(high_var_mat)), method = "complete")
clusters <- cutree(hc, k = 3) ## 3 clusters

df_pca <- data.frame(PC1 = pca_res$x[, 1], PC2 = pca_res$x[, 2], Cluster = factor(clusters)) ## rows are samples


ggplot(df_pca, aes(x = PC1, y = PC2, color = Cluster)) +
  geom_point(size = 3, alpha = 0.5) + 
  scale_color_manual(values = hue_pal()(length(unique(clusters)))) +
  theme_classic() + 
  labs(x = paste0("PC1: ", round(explained_variance[1], 1), "% variance"), 
       y = paste0("PC2: ", round(explained_variance[2], 1), "% variance"))
```

PCA and hierachical clustering in agreement in terms of groupings.

# Differential Expression Analysis

```{r}
# cluster the samples using the top 500 variable genes into 2 clusters
# perform differential expression analysis between the 2 clusters using DESeq


high_var_genes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 500) # extract indices of top 500 variable genes 
high_var_mat <- assay(vsd)[high_var_genes,] # subset gene expression matrix with indices
hc <- hclust(dist(t(high_var_mat)), method = "complete") # create dendrogram based on complete linkage
clusters <- cutree(hc, k = 2) # cut the dendrogram at level of 2 clusters


ccle_meta_subset$Cluster <- factor(clusters) # add a column to ccle_meta_subset labeling each sample with its assigned cluster

dds <- DESeqDataSetFromMatrix(countData = data.frame(ccle_counts_subset),
                              colData = ccle_meta_subset,
                              design = ~ Cluster) ## design parameter is now cluster


dds <- dds[rowSums(counts(dds)) >= 100,] ## filter out extremely lowly expressed genes
dds <- DESeq(dds) # do the differential expression analysis

res <- results(dds) ## get the results of the differential expression analysis
print(dim(res))
print(head(res, 20))
```

```{r}
summary(res)
```

```{r}
res <- res %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "gene_symbol") 
```

```{r}
res <- res %>% 
  filter(!is.na(padj)) ## filter out NAs in padj column
```

```{r}
low_p <- res[res$padj < 0.001, ] ## extract only the genes with an adjusted p value of < 0.001
ordered <- order(abs(low_p$log2FoldChange), decreasing = TRUE) # order genes by absolute value of log2 fold change
res_reordered <- low_p[ordered, ]
print(head(res_reordered, 30))
```

Top 3 DEGs in cluster 2: LYZ, REG4, MUC17 Top 3 DEGs in cluster 1:
MAGEA10, NPFFR2, BASP1

```{r}

top_6_DEGs <- c("LYZ", "REG4", "MUC17","MAGEA10", "NPFFR2", "BASP1")

# redo pca, create a dataframe with 2 columns containing the top 2 PCs
variances <- rowVars(assay(vsd))
top_var_genes <- order(variances, decreasing = TRUE)[1:500]
high_var_mat <- assay(vsd)[top_var_genes,]
pca_res <- prcomp(t(high_var_mat), center = TRUE, scale = FALSE)
explained_variance <- pca_res$sdev^2 / sum(pca_res$sdev^2) * 100
df_pca <- data.frame(PC1 = pca_res$x[, 1], PC2 = pca_res$x[, 2])

# add the top 6 DEG gene counts to the dataframe
PC1_PC2_top6_genes <- df_pca

for (gene in top_6_DEGs){
  PC1_PC2_top6_genes[[gene]] <- assay(vsd)[gene, ]
}

## pivot the table so that it is in long format
PC1_PC2_top6_genes_longer <- PC1_PC2_top6_genes |>
  mutate(Sample = rownames(PC1_PC2_top6_genes)) |>      # make row names a column
  pivot_longer(cols = LYZ:BASP1,               # select gene columns
               names_to = "Gene",
               values_to = "Expression")

# visualize the expression of the top 6 DEGs on PCA plots of the samples
ggplot(PC1_PC2_top6_genes_longer, aes(x = PC1, y = PC2, color = Expression)) +
  geom_point(size = 2, alpha = 0.5) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_classic() +
  labs(x = paste0("PC1: ", round(explained_variance[1], 1), "% variance"), 
       y = paste0("PC2: ", round(explained_variance[2], 1), "% variance"),
       color = "Expression (VST)") +
  facet_wrap(~ Gene)
```

There is evidence of differential expression of these genes along the
first principle component. As expected, 3 are highly expressed on the
"right" side of the axis, and 3 are highly expressed on the "left" side
of the axis, since I chose the top 3 DEGs in both clusters.

```{r}

top_ten_positive <- head(res_reordered[order(res_reordered$log2FoldChange, decreasing = TRUE),],10)$gene
top_ten_negative <- tail(res_reordered[order(res_reordered$log2FoldChange, decreasing = TRUE), ],10)$gene

top_genes <- c(top_ten_positive, top_ten_negative)

EnhancedVolcano(res_reordered,
  x = "log2FoldChange",    
  y = "padj",           
  lab = ifelse(res_reordered$gene %in% top_genes, res_reordered$gene, ""), ## 
  pCutoff = 0.00001,                              
  FCcutoff = 2,               
  labSize = 3,                
  title = "Cluster 1 vs Cluster 2 DEGs",
  subtitle = "",
  caption = "",
  pointSize = 1.0,   # smaller points
  colAlpha = 0.3,     # semi-transparent points
  drawConnectors = TRUE,
  widthConnectors = 0.5,
  max.overlaps = Inf
) +
  theme_classic() +
  theme(legend.position = "none")
```

```{r}
print(top_ten_positive)
print(top_ten_negative)
```

# Interpretation

cluster 2 genes: more associated with epithelium than cluster 1 genes -
Epithelial / mucosal lineage MUC17, MUC5AC: mucins, goblet/secretory
epithelial markers CEACAM5: epithelial adhesion molecule AGR3: secretory
cell-associated AQP5: water channel in mucosal secretory epithelia
SLC44A4: transporter enriched in intestinal epithelial cells - Goblet /
secretory / absorptive intestinal epithelial markers REG4: secretory
epithelial lineage LYZ: Paneth-like / innate defense

Cluster 1 genes: More heterogeneous, not strongly epithelial-specific
GHR: growth hormone receptor ADAMTS12: extracellular-matrix remodeling
enzyme GPC6: glypican (cell signaling, mesenchyme-related) EGFLAM:
adhesion/extracellular matrix KANK4: cytoskeletal regulation BASP1:
neuronal differentiation–related NPFFR2: neuropeptide receptor
MAGEA4/10: cancer/testis antigen family

# sessionInfo()

```{r}
sessionInfo()
```
